pkg/expression/util.go

```diff
@@ -553,6 +553,13 @@ func ColumnSubstituteImpl(ctx BuildContext, expr Expression, schema *Schema, new
 				var e Expression
 				var err error
 				if v.FuncName.L == ast.Cast {
+					// If the newArg is a ScalarFunction(cast), BuildCastFunctionWithCheck will modify the newArg.RetType,
+					// So we need to deep copy RetType.
+					// TODO: Expression interface needs a deep copy method.
+					if newArgFunc, ok := newArg.(*ScalarFunction); ok {
+						newArgFunc.RetType = newArgFunc.RetType.DeepCopy()
```

Comment:
Only `RetType` is deep copied here, there maybe new problem if other members of newArg was modified.
So we need a `DeepCopy` method in Expression interface.
But here I think only deep copy RetType is enough to fix this issue.